Extensions/GDevelopAuthentication/gdevelopauthenticationtools.ts

```diff
@@ -0,0 +1,865 @@
+      runtimeScene: gdjs.RuntimeScene
+    ) {
+      if (_authenticationBanner) {
+        // Banner already displayed, ensure it's visible.
+        _authenticationBanner.style.opacity = '1';
+        return;
+      }
+      if (!_checkedLocalStorage) {
+        fetchAuthenticatedUserFromLocalStorage();
+      }
+
+      const domElementContainer = runtimeScene
+        .getGame()
+        .getRenderer()
+        .getDomElementContainer();
+      if (!domElementContainer) {
+        handleAuthenticationError(
+          runtimeScene,
+          "The div element covering the game couldn't be found, the authentication banner cannot be displayed."
+        );
+        return;
+      }
+      if (_userToken && _username !== null) {
+        // If the user is already authenticated, we display the corresponding banner.
+        _authenticationBanner = computeAuthenticatedBanner(
+          runtimeScene,
+          _username
+        );
+        domElementContainer.appendChild(_authenticationBanner);
+        return;
+      } else {
+        // If the user is not authenticated, we display the corresponding banner.
+        _authenticationBanner = computeNotAuthenticatedBanner(runtimeScene);
+        domElementContainer.appendChild(_authenticationBanner);
+      }
+    };
+
+    /**
+     * Action to display the authentication window to the user.
+     */
+    export const displayAuthenticationWindow = function (
+      runtimeScene: gdjs.RuntimeScene
+    ) {
+      const _gameId = gdjs.projectData.properties.projectUuid;
+      const targetUrl = `https://liluo.io/oauth?gameId=${_gameId}&dev=true`;
+
+      startAuthenticationWindowTimeout(runtimeScene);
+
+      // Listen to messages posted by the authentication window, so that we can
+      // know when the user is authenticated.
+      _authenticationMessageCallback = (event: MessageEvent) => {
+        receiveMessageFromAuthenticationWindow(runtimeScene, event, {
+          checkOrigin: true,
+        });
+      };
+      window.addEventListener('message', _authenticationMessageCallback, true);
+
+      // Create the authentication container.
+      const domElementContainer = runtimeScene
+        .getGame()
+        .getRenderer()
+        .getDomElementContainer();
+      if (!domElementContainer) {
+        handleAuthenticationError(
+          runtimeScene,
+          "The div element covering the game couldn't be found, the authentication window cannot be displayed."
+        );
+        return;
+      }
+      const authenticationContainer = computeAuthenticationContainer(
+        runtimeScene
+      );
+      domElementContainer.appendChild(authenticationContainer);
+
+      // Based on which platform the game is running, we open the authentication window
+      // Inside the container, or in a new tab.
+      const electron = runtimeScene.getGame().getRenderer().getElectron();
+      if (electron) {
+        // If we're on Electron, use an iframe.
```

Comment:
Why use an iframe on electron? Users are likely to already be logged into their liluo account and link it from their default web browser. It is also a bad security practice as it teaches the users to enter their credentials in untrusted contexts instead of exclusively on the liluo site, and an iframe could most likely be read using a JS event to steal user's passwords :/ 

If it's because you cannot do window messaging when opening in another browser, what multiple other applications I've seen do is to register a custom protocol on the machine and redirect using it from the authentication page to give the token back to the electron application.

Reply:
This is indeed the problem (getting users to write credentials in untrusted contexts). And the issue with the browser is indeed getting a message back to the app.
Do you have an idea of how to do it with a browser?
I'm not sure how easy though it is to register a custom protocol.
Another way would be to do polling/long lived connection with the backend using a random token, but also unsure how robust/scalable it is

Reply:
Instead of polling, we could also use web socket to notify the game from the backend (by opening a web socket connection with a random token, open Liluo login in the browser with this token and send to the backend the JWT when the user has logged in. The backend can then notify the web socket connection.

We just have to prepare a message to tell the user to go back to the game.

Reply:
> Another way would be to do polling/long lived connection with the backend using a random token, but also unsure how robust/scalable it is

The AWS API gateway supports WebSockets, so you could do it kind of like this:
1. Connect over WebSocket to an OAuth API, and send an authentication request message
2. That triggers a lambda function via API gateway, send back the WebSocket connection ID
3. Open the Liluo oauth page with the WebSocket connection ID as get parameter
4. Once logged in (or the log in is cancelled), call a lambda function, send back the authentication result to the WebSocket that initiated the log in using the ID passed in as get parameter

Since it'd work on top of AWS for the unique ID generation & creation of long lived connections, and on top of lambdas, it should be pretty scalable I think. 

> I'm not sure how easy though it is to register a custom protocol.

It is pretty easy, but I think that a long lived connection to the backend should be more scalable, less error prone (what if two GDevelop games register the same protocol?) and reduce code duplication between Cordova and Electron

Reply:
Those are indeed good ideas 👍 
Completely agree with the fact that it's not good security practices.

As this extension is still experimental, it will not be widely used.
Also, the main usage will be on web to start with, as we gather feedback on usability.
So I'm suggesting to release is as is, and work on an updated version as soon as it is released.

Reply:
update: we've implemented a websocket connection when the application runs on Electron :)